### PR TITLE
(Don Syme) Added a patch for use support for IDisposable in cloud exp…

### PR DESCRIPTION
**Pull request by jasonweiyi**
…ressions.

### CODE REVIEW

**Comment by eiriktsarpalis** [comment]
The omission of IDisposable use bindings in cloud builders is intentional since most instances found in .NET are not serializable. So it's very easy to get unexpected runtime errors like this one:
```fsharp
cloud {
    use m = new System.IO.MemoryStream()
    let! result = Cloud.Parallel [ cloud { return m.Length } ]
    return result.[0]
}
```
Of course, removing use bindings per se do not cancel the possibility of such an error, however our goal here is to point out to people that they should really be using `local` workflows in such scenaria. These are guaranteed to be executed within the boundaries of a single address space, so you will not be getting any serialization-related errors.
```fsharp
local {
    use m = new System.IO.MemoryStream()
    let! result = Local.Parallel [ cloud { return m.Length } ]
    return result.[0]
}
```

/cc @dsyme

**Comment by dsyme** [comment]
Hmm...

* Is there an example that doesn't schedule a nested ``cloud { ... }`` block?  i.e. do any of the cloud.* builder methods result in implicit serialization/rescheduling of a computation fragment?

* If the goal is to encourage the use of ``local { ... }`` then an overload with an error or warning pointing in that direction would be best.  But I'm still sceptical that ``local { ... }`` is learnable for many users (see other thread on this).

**Comment by eiriktsarpalis** [comment]
Not necessarily, they would have to incorporate calls to `Cloud.Parallel/Choice/StartAsTask` or return the nonserializable value as a top-level result. Adding a compiler warning is probably a good compromise, we already do this in for loops and while loops with `cloud` anyway.

See: https://github.com/mbraceproject/MBrace.Core/blob/master/src/MBrace.Core/Continuation/Builders.fs#L333

**Comment by dsyme** [comment]
[ As an aside, I'm a bit unconvinced by the restriction on "while" loops - an actor uses a while loop, and there's no fundamental reason that MBrace cloud tasks can't be long running.... ]

**Comment by palladin** [comment]
The idea of "restricting" the use of imperative while loops is to discourage the use of ref-cell mutation which is problematic in the presence of arbitary serialization of values.

**Comment by dsyme** [comment]
The problem with this kind of restriction is that it inhibits transitions from normal F# code to ``cloud { ... }`` code.   I get why you're doing it, but 80-90% of the uses of MBrace won't have any nested ``cloud { ... }`` computations at all - and people who use MBrace really just have to learn to assess the set of captured variables of any scheduled cloud block for serializability. Trying to stop the problem at "use" doesn't address it in any fundamental way.  Indeed, when we rewrote the "use" as a "try-finally" then everything worked OK.

Basically, if rewriting to try/finally works, then use should work.

**Comment by dsyme** [comment]
Re imperative loops - I assume the following  code is always ok, is it (assuming no scheduling of nested ``cloud { ... }`` computations?

    cloud { let x = ref ... in  ... in while ... do ... }

That is, it is only on explicit operations like Cloud.Parallel that cloud blocks are serialized & scheduled.

**Comment by palladin** [comment]
I completely agree with "Trying to stop the problem at "use" doesn't address it in any fundamental way"

**Comment by palladin** [comment]
Sometimes it is tricky to track when serialization/scheduling is going to happen... as an example
```
cloud {
...
let! v = if ... then Cloud.Parallel [|...|] else cloud { return [|...|] }
...
}
```
and the idea behind local {}  is to have strong guaranties about the visibility of serialization/scheduling effects.

**Comment by dsyme** [comment]
I see, so the entire continuation is serialized, so

```
cloud {
    let v = ref 0 // or some non-serializable thing
    let! x = SomethingThatMayUseCloudParallel()
    ... reference to v ...
}
```

results in the possibility of invalid serialization of the continuation.

[ In this setting I still like the ``cloud0`` name - explained as "a `cloud { ... }` that starts precisely zero nested tasks, and no part of which gets serialized, and which can consequently use shared memory and unserializable in-memory objects safely" ]

**Comment by palladin** [comment]
I was wondering what that ```cloud0``` suffix zero means...

**Comment by palladin** [comment]
" a cloud { ... } that starts precisely zero nested tasks, and no part of which gets serialized, and which can consequently use shared memory and unserializable in-memory objects safely" it is exactly the motivation behind local {}

**Comment by eiriktsarpalis** [comment]
@dsyme We decided to go with the compiler warning in while loops after discovering that novice users -particularly ones lacking an FP background- would tend to use while loops with ref cells for iterative algorithms, resulting in instant confusion. This is trying to promote the idea that recursion is the natural pattern for distributed, iterative computation. 

While actors typically spend their lifetime in a single process, a cloud while loop may migrate the execution state to different machine(s) on each iteration. In that sense, an actor corresponds to a `local`/`cloud0` while loop.

**Comment by eiriktsarpalis** [comment]
I'm merging this, and adding a compiler warning to it.
